pep-0639.rst

```diff
@@ -1,543 +1,2221 @@
-  reference case for each SPDX license identifier and uppercase for the AND, OR
-  and WITH keywords.
 
-- SHOULD report an informational warning if normalization process results in
-  changes to the ``License`` field contents.
+Deprecate license classifiers
+'''''''''''''''''''''''''''''
 
-License expression examples::
+Using license `classifiers <#classifiers_>`_ in the ``Classifier`` field
+(described in PEP 301) is deprecated and replaced by the more precise
+``License-Expression`` field.
 
-    License: MIT
+If the ``License-Expression`` field is present, build tools SHOULD and
+publishing tools MUST raise an error if one or more license classifiers
+is included in a ``Classifier`` field, and MUST NOT add
+such classifiers themselves.
 
-    License: BSD-3-Clause
+Otherwise, if this field contains a license classifier, build tools MAY
+and publishing tools SHOULD issue a warning informing users such classifiers
+are deprecated, and recommending ``License-Expression`` instead.
+For compatibility with existing publishing and installation processes,
+the presence of license classifiers SHOULD NOT raise an error unless
+``License-Expression`` is also provided.
 
-    License: MIT OR GPL-2.0-or-later OR (FSFUL AND BSD-2-Clause)
+For all newly-uploaded distributions that include a
+``License-Expression`` field, the `Python Package Index (PyPI) <#pypi_>`_ MUST
+reject any that also specify any license classifiers.
 
-    License: GPL-3.0-only WITH Classpath-Exception-2.0 OR BSD-3-Clause
+New license classifiers MUST NOT be `added to PyPI <#classifiersrepo_>`_;
+users needing them SHOULD use the ``License-Expression`` field instead.
+Along with the ``License`` field, license classifiers may be removed from a
+new version of the specification in a future PEP.
 
-    License: This software may only be obtained by sending the
-            author a postcard, and then the user promises not
-            to redistribute it.
 
-    License: LicenseRef-Proprietary AND LicenseRef-Public-Domain
+Project source metadata
+-----------------------
 
+As originally introduced in PEP 621, the
+`PyPA Declaring Project Metadata specification <#pep621spec_>`_
+defines how to declare a project's source
+metadata in a ``[project]`` table in the ``pyproject.toml`` file for
+build tools to consume and output distribution core metadata.
 
-Classifier (multiple use)
-:::::::::::::::::::::::::
+This PEP `adds <Add license-expression key_>`_ the ``license-expression`` key,
+`adds <Add license-files key_>`_ the ``license-files`` key and
+`deprecates <Deprecate license key_>`_ the ``license`` key.
 
-Each entry is a string giving a single classification value for the
-distribution. Classifiers are described in PEP 301.
 
-Examples::
+Add ``license-expression`` key
+''''''''''''''''''''''''''''''
 
-    Classifier: Development Status :: 4 - Beta
-    Classifier: Environment :: Console (Text Based)
+A new ``license-expression`` key is added to the ``project`` table, which has
```

Comment:
I'd really strongly prefer that we use the existing `license` key instead.

The plain text form for the `license` key was intentionally kept unused for this specific usecase. From https://python.github.io/peps/pep-0621/#license:

> A practical string value for the license key has been purposefully left out to allow for a future PEP to specify support for SPDX [6] expressions (the same logic applies to any sort of “type” field specifying what license the file or text represents).

There's absolutely no reason to introduce a new key here; and certainly no reason to mirror key names under `project` with what's in the `METADATA` file.

Reply:
I agree - this is what the `license` key was intended for, and we should use it as defined.

But this discussion should happen on Discourse, as noted below.

Reply:
> The plain text form for the license key was intentionally kept unused for this specific usecase. From https://python.github.io/peps/pep-0621/#license:

Just to note once again, I'm well aware of this and it was my initial preference too, and what I initially implemented in the PEP, as I discuss in detail under the Rejected Ideas section.

> There's absolutely no reason to introduce a new key here; and certainly no reason to mirror key names under project with what's in the METADATA file.

Actually, there are, as I took the time to describe in detail under the Rejected Ideas section, and as recently discussed at length on this PEP's Discourse thread (the most important one being the hard conflict with the use of the `dynamic` key, and reader and author confusion and ambiguity with other tool formats). I revised it now to address the points raised more explicitly and focus more on those core issues while minimizing the rest.

I'm happy to discuss this further and hear your ideas for other ways around these issues on the Discourse thread.

> I agree - this is what the license key was intended for, and we should use it as defined.

While reserving the string value for _potential_ not-yet-specified future use is normative, the multiple suggestions for how more precise license metadata might be implemented does not appear to be, being contained in an explanatory note, specified as an envisioned future use, and with multiple options specified; and it does not consider the practical implementation complexities and interactions with the `dynamic` field that I've spent much time analyzing in detail here, and which we can discuss further on the thread.

> But this discussion should happen on Discourse, as noted below.

Agreed; for the record, I tried to politely steer previous substantive discussion to the Discourse thread the first time this was brought up, but didn't entirely succeed, sorry.

Reply:
Yea, here's the thread: https://discuss.python.org/t/12622